contracts/collection/templates/NFTCollection.sol

```diff
@@ -0,0 +1,472 @@
+        // The maximum number of tokens that can be minted in this collection.
+        uint256 maxSupply;
+        // The number of free token mints reserved for the contract owner
+        uint256 reservedSupply;
+        // Minting price per token.
+        uint256 mintPrice;
+        // The maximum number of tokens the user can mint per transaction.
+        uint256 tokensPerMint;
+        // Treasury address is the address where minting fees can be withdrawn to.
+        // Use `withdrawFees()` to transfer the entire contract balance to the treasury address.
+        address payable treasuryAddress;
+    }
+
+    /// Updatable by admins and owner
+    struct RuntimeConfig {
+        // Metadata base URI for tokens, NFTs minted in this contract will have metadata URI of `baseURI` + `tokenID`.
+        // Set this to reveal token metadata.
+        string baseURI;
+        // If true, the base URI of the NFTs minted in the specified contract can be updated after minting (token URIs
+        // are not frozen on the contract level). This is useful for revealing NFTs after the drop. If false, all the
+        // NFTs minted in this contract are frozen by default which means token URIs are non-updatable.
+        bool metadataUpdatable;
+        // Starting timestamp for public minting.
+        uint256 publicMintStart;
+        // Starting timestamp for whitelisted/presale minting.
+        uint256 presaleMintStart;
+        // Pre-reveal token URI for placholder metadata. This will be returned for all token IDs until a `baseURI`
+        // has been set.
+        string prerevealTokenURI;
+        // Root of the Merkle tree of whitelisted addresses. This is used to check if a wallet has been whitelisted
+        // for presale minting.
+        bytes32 presaleMerkleRoot;
+        // Secondary market royalties in basis points (100 bps = 1%)
+        uint256 royaltiesBps;
+        // Address for royalties
+        address royaltiesAddress;
+    }
+
+    struct ContractInfo {
+        uint256 version;
+        DeploymentConfig deploymentConfig;
+        RuntimeConfig runtimeConfig;
+    }
+
+    event OwnershipTransferred(
+        address indexed previousOwner,
+        address indexed newOwner
+    );
+
+    /*************
+     * Constants *
+     *************/
+
+    /// Contract version, semver-style uint X_YY_ZZ
+    uint256 public constant VERSION = 1_02_00;
+
+    /// Admin role
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+
+    // Basis for calculating royalties.
+    // This has to be 10k for royaltiesBps to be in basis points.
+    uint16 constant ROYALTIES_BASIS = 10000;
+
+    /********************
+     * Public variables *
+     ********************/
+
+    /// The number of currently minted tokens
+    /// @dev Managed by the contract
+    uint256 public totalSupply;
+
+    /***************************
+     * Contract initialization *
+     ***************************/
+
+    constructor() ERC721("", "") {
```

Comment:
btw dont we need name and symbol in ERC721 constructor? :O

Reply:
Not really, we override `name()` and `symbol()` later in the `Internals` section so the contract still has them implemented properly